File: rxbinding/src/main/java/com/jakewharton/rxbinding/widget/RxPopupMenu.java
package com.jakewharton.rxbinding.widget;

import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.view.MenuItem;
import android.widget.PopupMenu;
import rx.Observable;

import static com.jakewharton.rxbinding.internal.Preconditions.checkNotNull;

/**
 * Static factory methods for creating {@linkplain Observable observables} for {@link PopupMenu}.
 */
public final class RxPopupMenu {
  /**
   * Create an observable which emits the clicked item in {@code view}'s menu.
   * <p>
   * <em>Warning:</em> The created observable keeps a strong reference to {@code view}.
   * Unsubscribe to free this reference.
   * <p>
   * <em>Warning:</em> The created observable uses {@link PopupMenu#setOnMenuItemClickListener}
   * to observe dismiss change. Only one observable can be used for a view at a time.
   */
  @CheckResult @NonNull
  public static Observable<MenuItem> itemClicks(@NonNull PopupMenu view) {
    checkNotNull(view, "view == null");
    return Observable.create(new PopupMenuItemClickOnSubscribe(view));
  }

  /**
   * Create an observable which emits on {@code view} dismiss events. The emitted value is
   * unspecified and should only be used as notification.
   * <p>
   * <em>Warning:</em> The created observable keeps a strong reference to {@code view}.
   * Unsubscribe to free this reference.
   * <p>
   * <em>Warning:</em> The created observable uses {@link PopupMenu#setOnDismissListener} to
   * observe dismiss change. Only one observable can be used for a view at a time.
   */
  @CheckResult @NonNull
  public static Observable<Void> dismisses(@NonNull PopupMenu view) {
    checkNotNull(view, "view == null");
    return Observable.create(new PopupMenuDismissOnSubscribe(view));
  }

  private RxPopupMenu() {
    throw new AssertionError("No instances.");
  }
}

File: rxbinding-appcompat-v7-kotlin/src/main/kotlin/com/jakewharton/rxbinding/support/v7/widget/RxPopupMenu.kt
package com.jakewharton.rxbinding.support.v7.widget

import android.support.v7.widget.PopupMenu
import android.view.MenuItem
import rx.Observable

/**
 * Create an observable which emits the clicked item in `view`'s menu.
 *
 * *Warning:* The created observable keeps a strong reference to `view`.
 * Unsubscribe to free this reference.
 *
 * *Warning:* The created observable uses [PopupMenu.setOnMenuItemClickListener]
 * to observe dismiss change. Only one observable can be used for a view at a time.
 */
public inline fun PopupMenu.itemClicks(): Observable<MenuItem> = RxPopupMenu.itemClicks(this)

/**
 * Create an observable which emits on `view` dismiss events. The emitted value is
 * unspecified and should only be used as notification.
 *
 * *Warning:* The created observable keeps a strong reference to `view`.
 * Unsubscribe to free this reference.
 *
 * *Warning:* The created observable uses [PopupMenu.setOnDismissListener] to
 * observe dismiss change. Only one observable can be used for a view at a time.
 */
public inline fun PopupMenu.dismisses(): Observable<Unit> = RxPopupMenu.dismisses(this).map { Unit }
